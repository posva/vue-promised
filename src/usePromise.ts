import {
  ref,
  watch,
  Ref,
  computed,
  ComputedRef,
  unref,
  // TODO:
  // toValue
  // MaybeRefOrGetter
} from 'vue-demi'

export type MaybeRef<T = unknown> = T | Ref<T>
export type MaybeRefOrGetter<T = unknown> = MaybeRef<T> | (() => T)
function toValue<T>(source: MaybeRefOrGetter<T>): T {
  // @ts-expect-error: source not callable
  return typeof source === 'function' ? source() : unref(source)
}

/**
 * Returns the state of a Promise and observes the Promise if it's a Ref to
 * automatically update the state
 *
 * @param promise - Ref of a Promise or raw Promise
 * @param pendingDelay - optional delay to wait before displaying pending
 */
export function usePromise<T = unknown>(
  promise: MaybeRef<Promise<T> | null | undefined>,
  pendingDelay: MaybeRef<number | string> = 200
): UsePromiseResult<T> {
  const isRejected = ref(false)
  const isResolved = ref(false)
  const isPending = computed(() => !isRejected.value && !isResolved.value)
  const isDelayElapsed = ref(false)
  const error = ref<Error | undefined | null>()
  const data = ref<T | null | undefined>()

  let timerId: ReturnType<typeof setTimeout> | undefined | null

  watch(
    () => toValue(promise),
    (newPromise) => {
      isRejected.value = false
      isResolved.value = false
      error.value = null
      if (!newPromise) {
        data.value = null
        if (timerId) clearTimeout(timerId)
        timerId = null
        return
      }

      const pendingDelayNumber = Number(toValue(pendingDelay)) || 0
      if (pendingDelayNumber > 0) {
        isDelayElapsed.value = false
        if (timerId) clearTimeout(timerId)
        timerId = setTimeout(() => {
          isDelayElapsed.value = true
        }, pendingDelayNumber)
      } else {
        isDelayElapsed.value = true
      }

      newPromise
        .then((newData) => {
          // ensure we are dealing with the same promise
          if (newPromise === toValue(promise)) {
            data.value = newData
            isResolved.value = true
          }
        })
        .catch((err) => {
          // ensure we are dealing with the same promise
          if (newPromise === toValue(promise)) {
            error.value = err
            isRejected.value = true
          }
        })
    },
    { immediate: true }
  )

  return { isPending, isRejected, isResolved, isDelayElapsed, error, data }
}

/**
 * Return type of `usePromise()`
 */
export interface UsePromiseResult<T = unknown> {
  isPending: ComputedRef<boolean>
  isResolved: Ref<boolean>
  isRejected: Ref<boolean>
  isDelayElapsed: Ref<boolean>
  error: Ref<Error | undefined | null>
  data: Ref<T | undefined | null>
}
