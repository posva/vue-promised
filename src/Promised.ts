import { watch } from 'vue'
import {
  defineComponent,
  isVue3,
  PropType,
  reactive,
  toRefs,
  warn,
  AllowedComponentProps,
  ComponentCustomProps,
  VNodeProps,
  VNode,
} from 'vue-demi'
import { usePromise, UsePromiseResult } from './usePromise'

export const PromisedImpl = /*#__PURE__*/ defineComponent({
  name: 'Promised',
  props: {
    promise: {} as PropType<Promise<unknown> | null | undefined>,
    // validator: p =>
    //   p && typeof (p as any).then === 'function' && typeof (p as any).catch === 'function',

    pendingDelay: {
      type: [Number, String],
      default: 200,
    },
  },
  emits: ['resolved', 'rejected', 'pending'],
  setup(props, { slots, emit, attrs }) {
    const propsAsRefs = toRefs(props)
    const promiseState = reactive(
      usePromise(propsAsRefs.promise, propsAsRefs.pendingDelay)
    )

    watch(promiseState, (promiseState) =>
      promiseState.isRejected
        ? emit('rejected', promiseState.error)
        : !promiseState.isPending
        ? emit('resolved', promiseState.data)
        : promiseState.isDelayElapsed
        ? emit('pending', promiseState.data)
        : null
    )

    return () => {
      if ('combined' in slots) {
        return slots.combined!(promiseState)
      }

      const [slotName, slotData] = promiseState.isRejected
        ? ['rejected', promiseState.error]
        : !promiseState.isPending
        ? ['default', promiseState.data]
        : promiseState.isDelayElapsed
        ? ['pending', promiseState.data]
        : [null]

      if (__DEV__ && slotName && !slots[slotName]) {
        ;(isVue3 ? warn : console.warn)(
          `(vue-promised) Missing slot "${slotName}" in Promised component. Pass an empty "${slotName}" slot or use the "combined" slot to remove this warning. This will fail in production.`
        )
        return null
      }

      return slotName && slots[slotName]!(slotData)
    }
  },
})

export const Promised = PromisedImpl as unknown as {
  new (): {
    $props: AllowedComponentProps &
      ComponentCustomProps &
      VNodeProps & { promise: Promise<unknown> }
  }

  $slots: {
    default: (data: unknown) => VNode[]
    rejected: (error: Error) => VNode[]
    pending: (error: unknown) => VNode[]

    combined: (arg: UsePromiseResult) => VNode[]
  }
}
