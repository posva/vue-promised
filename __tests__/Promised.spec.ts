// @vitest-environment jsdom
/* eslint-disable no-unused-vars */
import { mount } from '@vue/test-utils'
import { PromisedImpl as Promised } from '../src/Promised'
import fakePromise from 'faked-promise'
import { h } from 'vue'
import {
  describe,
  it,
  expect,
  beforeEach,
  afterEach,
  afterAll,
  vi,
} from 'vitest'
import { mockWarn } from './vitest-mock-warn'

// keep a real setTimeout
const timeout = setTimeout
const tick = () => new Promise((resolve) => timeout(resolve, 0))

describe('Promised', () => {
  function factory(propsData: any = undefined, slots: any = undefined) {
    const [promise, resolve, reject] = fakePromise<any>()
    const wrapper = mount(Promised, {
      propsData: { promise, pendingDelay: 0, ...propsData },
      slots: {
        pending: (oldData) => h('span', 'pending: ' + oldData),
        default: (data) => h('span', {}, data),
        rejected: (error) =>
          h('span', { class: 'error' }, error && error.message),
        ...slots,
      },
    })

    return { wrapper, promise, resolve, reject }
  }

  describe('three slots', () => {
    it('displays nothing with no promise', () => {
      const { wrapper } = factory({ promise: null })
      expect(wrapper.text()).toBe('')
    })

    it('displays pending', () => {
      const { wrapper } = factory()
      expect(wrapper.text()).toMatch('pending')
    })

    it('displays the resolved value once resolved', async () => {
      const { wrapper, resolve } = factory()
      resolve('foo')
      await tick()
      expect(wrapper.text()).toBe('foo')
    })

    it('contains previous data in pending slot', async () => {
      let { wrapper, promise, resolve } = factory()
      resolve('ok')
      await tick()
      expect(wrapper.text()).toBe('ok')
      // create a new promise
      ;[promise, resolve] = fakePromise()
      wrapper.setProps({ promise })
      await tick()
      resolve('okay')
      expect(wrapper.text()).toBe('pending: ok')
      await tick()
      expect(wrapper.text()).toBe('okay')
    })

    it('displays an error if rejected', async () => {
      let { wrapper, reject } = factory()
      reject(new Error('hello'))
      await tick()
      expect(wrapper.text()).toBe('hello')
    })

    it('should display rejected slot even if error is undefined', async () => {
      let { wrapper, reject } = factory()

      reject(undefined)
      await tick()
      expect(wrapper.find('.error').exists()).toBeTruthy()
    })

    it('cancels previous promise', async () => {
      let { wrapper, resolve } = factory()
      const other = fakePromise()
      wrapper.setProps({ promise: other[0] })
      resolve('foo')
      await tick()
      expect(wrapper.text()).toMatch('pending')
    })

    it('cancels previous rejected promise', async () => {
      let { wrapper, reject } = factory()
      const other = fakePromise()
      wrapper.setProps({ promise: other[0] })
      reject(new Error('failed'))
      await tick()
      expect(wrapper.text()).toMatch('pending')
    })
  })

  describe('three slots pendingDelay', () => {
    beforeEach(() => {
      vi.useFakeTimers()
      vi.spyOn(global, 'setTimeout')
      vi.spyOn(global, 'clearTimeout')
    })
    afterEach(() => {
      // @ts-expect-error: mocked
      setTimeout.mockClear()
      // @ts-expect-error: mocked
      clearTimeout.mockClear()
    })

    afterAll(() => {
      vi.useRealTimers()
    })

    it('displays nothing before the delay', async () => {
      let { wrapper } = factory({ pendingDelay: 300 })
      expect(wrapper.text()).toBe('')
      vi.runAllTimers()
      await wrapper.vm.$nextTick()
      expect(wrapper.text()).toMatch('pending')
    })

    it('custom pendingDelay', async () => {
      expect(setTimeout).toHaveBeenCalledTimes(0)
      let { wrapper, promise } = factory({ pendingDelay: 300 })
      expect(setTimeout).toHaveBeenCalledTimes(1)
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 300)
      ;[promise] = fakePromise()
      wrapper.setProps({ pendingDelay: 100, promise })
      await wrapper.vm.$nextTick()
      expect(setTimeout).toHaveBeenCalledTimes(2)
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 100)
    })

    it('cancels previous timeouts', async () => {
      let { wrapper, promise } = factory({ pendingDelay: 300 })
      expect(clearTimeout).not.toHaveBeenCalled()
      ;[promise] = fakePromise()
      wrapper.setProps({
        promise,
        pendingDelay: 100,
      })
      await wrapper.vm.$nextTick()
      expect(clearTimeout).toHaveBeenCalled()
    })

    it('cancels timeout when promise is set to null', async () => {
      let { wrapper } = factory({ pendingDelay: 300 })
      expect(setTimeout).toHaveBeenCalledTimes(1)
      wrapper.setProps({
        promise: null,
      })
      await wrapper.vm.$nextTick()
      expect(clearTimeout).toHaveBeenCalledTimes(1)
    })
  })

  describe('combined slot', () => {
    function factory(propsData: any = undefined, slots: any = undefined) {
      const [promise, resolve, reject] = fakePromise<any>()
      const wrapper = mount(Promised, {
        propsData: { promise, pendingDelay: 0, ...propsData },
        slots: {
          combined: (state) => h('span', JSON.stringify(state)),
          ...slots,
        },
      })

      return { wrapper, promise, resolve, reject }
    }

    it('displays the data', async () => {
      const { wrapper, resolve } = factory()

      expect(wrapper.html()).toMatchSnapshot()

      resolve('foo')
      await tick()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('displays the error', async () => {
      const { wrapper, reject } = factory()

      expect(wrapper.html()).toMatchSnapshot()

      reject(new Error('fail'))
      await tick()

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  mockWarn()

  it('warns on missing slot', async () => {
    const [promise, _resolve, reject] = fakePromise<any>()
    mount(Promised, {
      propsData: { promise, pendingDelay: 0 },
      slots: {
        pending: (oldData) => h('span', 'pending: ' + oldData),
        default: (data) => h('span', {}, data),
        // rejected: (error) => h('span', { class: 'error' }, error.message),
      },
    })

    reject(new Error())
    await tick()

    expect(/Missing slot "rejected"/).toHaveBeenWarned()
  })
})
