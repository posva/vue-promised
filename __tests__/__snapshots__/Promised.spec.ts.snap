// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Promised > combined slot > displays the data 1`] = `"<span>{\\"isPending\\":true,\\"isRejected\\":false,\\"isResolved\\":false,\\"isDelayElapsed\\":true,\\"error\\":null}</span>"`;

exports[`Promised > combined slot > displays the data 2`] = `"<span>{\\"isPending\\":false,\\"isRejected\\":false,\\"isResolved\\":true,\\"isDelayElapsed\\":true,\\"error\\":null,\\"data\\":\\"foo\\"}</span>"`;

exports[`Promised > combined slot > displays the error 1`] = `"<span>{\\"isPending\\":true,\\"isRejected\\":false,\\"isResolved\\":false,\\"isDelayElapsed\\":true,\\"error\\":null}</span>"`;

exports[`Promised > combined slot > displays the error 2`] = `"<span>{\\"isPending\\":false,\\"isRejected\\":true,\\"isResolved\\":false,\\"isDelayElapsed\\":true,\\"error\\":{}}</span>"`;
